code/datums/personal_summary.dm

```diff
@@ -0,0 +1,96 @@
+///Global list of personal summaries by client key
+var/global/list/datum/personal_summary/personal_summary_by_client = list()
+
+///Combination of personal metacurrency windows, shown end-of-round
+/datum/personal_summary
+	// jobxp
+	var/current_job
+	var/current_level
+	var/earned_exp
+	var/level_exp
+	var/total_exp
+	var/next_level_exp
+	var/exp_earned
+
+	// spacebux
+	var/is_antagonist
+	var/is_part_time
+	var/is_escaped
+	var/is_pilot
+	var/base_wage
+	var/score_adjusted_wage
+	var/objective_completed_bonus
+	var/all_objectives_bonus
+	var/pilot_bonus
+	var/earned_spacebux
+	var/total_spacebux
+	var/held_item
+
+	var/personal_summary_data
+
+
+/datum/personal_summary/ui_state(mob/user)
+	return tgui_always_state.can_use_topic(src, user)
+
+/datum/personal_summary/ui_status(mob/user, datum/ui_state/state)
+	return tgui_always_state.can_use_topic(src, user)
+
+/datum/personal_summary/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "PersonalSummary")
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/datum/personal_summary/proc/generate_xp(key, mob/M)
+	SPAWN(0)
+		if(key in xp_archive)
+			var/list/keyList = xp_archive[key]
+			var/hasEntries = 0
+			for(var/job in keyList)
+				hasEntries = 1
+				src.exp_earned = TRUE
+				src.current_job = job
+				src.earned_exp = keyList[job]
+				src.total_exp = get_xp(key, job)
+
+				if (isnull(src.total_exp))
+					src.total_exp = src.earned_exp // for local dev servers where get_xp will always return null
+				src.current_level = LEVEL_FOR_XP(src.total_exp)
+				src.level_exp = src.total_exp - XP_FOR_LEVEL(src.current_level)
+				src.next_level_exp = XP_FOR_LEVEL(src.current_level  + 1) - XP_FOR_LEVEL(src.current_level)
+
+			if(!hasEntries)
+				src.exp_earned = FALSE
+
+		src.ui_interact(M)
```

Comment:
I guess this has to be here because of the spawn? I would be nice if it were called in `declare_completion()` right before adding the ability